ext/msgpack/packer.c

```diff
@@ -114,6 +114,40 @@ bool msgpack_packer_try_write_with_ext_type_lookup(msgpack_packer_t* pk, VALUE v
     }
 
     if(ext_flags & MSGPACK_EXT_RECURSIVE) {
+        // HACK: While we call the proc, the current pk->buffer won't be reachable
+        // as it will be stored on the stack.
+        // To ensure all the `mapped_string` reference in that buffer are properly
+        // marked and pined, we copy them all on the stack.
+        VALUE* mapped_strings = NULL;
+        size_t mapped_strings_count = 0;
+        msgpack_buffer_chunk_t* c = pk->buffer.head;
+        while (c != &pk->buffer.tail) {
+            if (c->mapped_string != NO_MAPPED_STRING) {
+                mapped_strings_count++;
+            }
+            c = c->next;
+        }
+        if (c->mapped_string != NO_MAPPED_STRING) {
+            mapped_strings_count++;
+        }
+
+        if (mapped_strings_count > 0) {
+            mapped_strings = ALLOCA_N(VALUE, mapped_strings_count);
```

Comment:
How large can `mapped_strings_count` get? I'm worried about potential stack overflows.

Reply:
Unfortunately, there isn't really a specific limit.

We could store them in an Array, but then they wouldn't be pinned.

Reply:
Yeah, I think you are right.